### PR TITLE
Ensure warp and teleport target position is free

### LINKED ISSUE #6338: Warp can still place you on a solid tile
The starting point (red) was used by the hero to warp. After using the warp spell the level went black.
black

Savegame:
[black.zip](https://github.com/diasurgical/devilutionX/files/12003467/black.zip)

When the hero starts moving the level goes back to normal.

### PR DESCRIPTION
**Pull request by obligaron**
Fixes #6338

Notes
- Warp didn't check target position is not free. This includes solid objects (source of #6338) but also true for monsters and players. So with warp you could place yourself on the same tile as a monster. :wink:
- Warp can only be casted by a player-"source". So simplified `AddWarp`. `ProcessTeleport` also doesn't check missile source, so it should be fine.
- Teleport (and phasing) position is checked on cast (`AddTeleport`, `AddPhasing`) but not validated again when the teleport is executed (`ProcessTeleport`). Normally this is not a problem, because `AddTeleport` and `ProcessTeleport` happens in the same game tick. But in multiplayer with two players both casting simultaneous on the same target tile this initial check is not sufficient. So another check in `ProcessTeleport` is needed. I noticed this kind of desync (two players on one tile) when watching some PvP players and I think this missing check could be the culprit.

### CODE REVIEW

**Review by AJenbo — APPROVED** [review]
Thanks for handling this. Hopefully people won't get pushed on to the trigger to often :D

File: Source/missiles.cpp [hunk 1]
```diff
@@ -1470,15 +1467,27 @@ void AddWarp(Missile &missile, AddMissileParameter & /*parameter*/)
...
+	std::optional<Point> teleportDestination = FindClosestValidPosition(
+	    [&player](Point target) {
+		    return PosOkPlayer(player, target);
+	    },
+	    tile, 0, 5);
+
+	if (teleportDestination) {
+		missile.position.tile = *teleportDestination;
+	} else {
+		// No valid teleport destiniation found
```
Comment (qndel) [inline Source/missiles.cpp:1487]:
```suggestion
		// No valid teleport destination found
```

**Comment by kphoenix137** [comment]
Nice! In PvP in the past I have noticed instances where the other player and I will end up on the same tile, and one of us ends up invisible until they start walking again.

**Comment by obligaron** [comment]
> Thanks for handling this. Hopefully people won't get pushed on to the trigger to often :D

i will look if I can add a trigger check. This would ensure that warp doesn't position the player directly on a trigger tile. 🙂

> Nice! In PvP in the past I have noticed instances where the other player and I will end up on the same tile, and one of us ends up invisible until they start walking again. Glad it's getting a fix

I hope this get fixed with this pr. PvP has so many crazy teleport and phasing action. 😁

File: Source/missiles.cpp (hunk 1 above)
Comment (obligaron) [reply Source/missiles.cpp:1487]: Fixed. Thanks 🙂

**Review by AJenbo — APPROVED** [review]

**Pull request merged**